### PR TITLE
System invisible fence radius

### PR DESCRIPTION
**Pull request by a-random-lemurian**
## Summary
System-specific invisible fences can now be set. This will allow plugins to have bigger systems that truly feel big, without having to worry about NPCs going in circles the moment they reach the invisible fence.

~~This is something I bodged together and it is possible that this is nonfunctional, so please help me spot the errors in my code.~~ It seems to work now.

## Testing Done
I messed around with the `"invisible fence"` value in my plugin. See the fruits of my labor in the Usage Examples section right below.

## Usage Examples
Look at these ships! They are trying to enter a wormhole that is 30,000 units from the center. Unfortunately the system fence is 10,000 units so the AI ends up fighting itself. One part is telling the ship to go to the wormhole while another is telling the ship to turn around. Also the ships will pile up over time and cause extreme framerate drops. This sucks!

image

And now let's see what happens if I set the `"invisible fence"` to 35,000 units.

- *edits plugin file, adds line `"invisible fence" 35000`*
- *reloads the game*

image
image
How wonderful! The ships are now flying all the way to the wormhole because now the system fence is 35,000 units from the center!

## Performance Impact
Unsure, but again letting ships pile up in big systems will cause some extreme drops in framerate.

Edit 1, 2022-11-25: Updated the description

### CODE REVIEW

File: source/System.h [hunk 1]
```diff
@@ -234,6 +236,9 @@ class System {
 
 	// Attributes, for use in location filters.
 	std::set<std::string> attributes;
+
+	// The "invisible fence" radius for this system.
+	double invisibleFenceRadius = 10000;
```
Comment (warp-core) [inline source/System.h:241]:
```suggestion
	double invisibleFenceRadius = 10000.;
```

**Review by warp-core — COMMENTED** [review]
I don't think AI is used statically, but that variable in the namespace is going to be static. That's not good.

**Comment by Koranir** [comment]
The variable MAX_DISTANCE_FROM_CENTER should be changed to something like centerMaxDistance and moved to the header as it is no longer a constant.

**Comment by Zitchas** [comment]
Being able to override the system fence on a per-system basis could be very useful even in vanilla. Thanks for working on this!

**Comment by a-random-lemurian** [comment]
> @Zitchas said:
> Being able to override the system fence on a per-system basis could be very useful even in vanilla. Thanks for working on this!

You're welcome. Now my plugin (and vanilla, and all other plugins) can have bigger systems without the need to worry about the ships going around in circles.

However, I'm not ready for this to be merged yet. I still have some things I'd like to discuss.

- Should we use `"invisible fence"` as the keyword to override the system fence or should we use a different word like `"fence"`?

**Comment by RisingLeaf** [comment]
The AI gas access to the system class, no need to store the radius in the AI class

**Comment by a-random-lemurian** [comment]
> The AI gas access to the system class, no need to store the radius in the AI class

refactor time

File: source/AI.cpp [hunk 2]
```diff
@@ -4065,3 +4066,16 @@ void AI::UpdateOrders(const Ship &ship)
...
+void AI::SetInvisibleFenceRadius(double radius)
+{
+	this->invisibleFenceRadius = radius;
+}
+
+
+
+double AI::GetInvisibleFenceRadius() const
+{
+	return this->invisibleFenceRadius;
```
Comment (RisingLeaf) [inline source/AI.cpp:4080]:
```suggestion
	return invisibleFenceRadius;
```

I mean this function will be uneccessary if you look at my last comment, but for the future, class members do not need to be accessed with this->

File: source/AI.cpp [hunk 3]
```diff
@@ -514,7 +513,7 @@ void AI::Step(const PlayerInfo &player, Command &activeCommands)
 			++it;
 	}
 	for(const auto &it : ships)
-		if(it->Position().Length() >= MAX_DISTANCE_FROM_CENTER)
+		if(it->Position().Length() >= GetInvisibleFenceRadius())
```
Comment (RisingLeaf) [inline source/AI.cpp:516]:
```suggestion
		if(it->Position().Length() >= ship.GetSystem()->InvisibleFenceRadius())
```

**Comment by a-random-lemurian** [comment]
Weird. This PR is segfaulting when I enter a system with a wormhole.

**Comment by wjp** [comment]
One potential source of segfaults here is when looping through the list of ships in `AI::Step()` some ships might have a null System (for carried drones/fighters).

**Comment by a-random-lemurian** [comment]
> @wjp said:
> One potential source of segfaults here is when looping through the list of ships in `AI::Step()` some ships might have a null System (for carried drones/fighters).

That was it all along! It works now. Let's make sure the tests succeed.

File: source/System.cpp [hunk 4]
```diff
@@ -371,6 +371,13 @@ void System::Load(const DataNode &node, Set<Planet> &planets)
 					grand.PrintTrace("Warning: Skipping unsupported departure distance limitation:");
 			}
 		}
+		else if(key == "invisible fence")
+		{
+			if(child.Size() >= 2)
+			{
+				invisibleFenceRadius = child.Value(1);
+			}
+		}
```
Comment (RisingLeaf) [inline source/System.cpp:380]:
```suggestion
		else if(key == "invisible fence" && child.Size() >= 2)
			invisibleFenceRadius = child.Value(1);
```

**Review by RisingLeaf — CHANGES_REQUESTED** [review]

**Review by RisingLeaf — APPROVED** [review]

File: source/AI.cpp [hunk 5]
```diff
@@ -514,11 +513,17 @@ void AI::Step(const PlayerInfo &player, Command &activeCommands)
...
+	{
+		// Skip carried fighters or drones.
+		if(!it->GetSystem())
+			continue;
+		else
+			if(it->Position().Length() >= it->GetSystem()->InvisibleFenceRadius())
+			{
+				int &value = fenceCount[&*it];
+				value = min(FENCE_MAX, value + FENCE_DECAY + 1);
+			}
```
Comment (Amazinite) [inline source/AI.cpp:525]:
```suggestion
		const System *system = it->GetActualSystem()
		if(system && it->Position().Length() >= system ->InvisibleFenceRadius())
		{
			int &value = fenceCount[&*it];
			value = min(FENCE_MAX, value + FENCE_DECAY + 1);
		}
```

File: source/System.cpp [hunk 6]
```diff
@@ -964,3 +966,10 @@ void System::Price::Update()
 {
 	price = base + static_cast<int>(-100. * erf(supply / LIMIT));
 }
+
+
+
+double System::InvisibleFenceRadius() const
+{
+	return this->invisibleFenceRadius;
```
Comment (Amazinite) [inline source/System.cpp:974]:
```suggestion
	return invisibleFenceRadius;
```

File: source/System.cpp [hunk 7]
```diff
@@ -964,3 +966,10 @@ void System::Price::Update()
 {
 	price = base + static_cast<int>(-100. * erf(supply / LIMIT));
 }
+
+
+
+double System::InvisibleFenceRadius() const
```
Comment (Amazinite) [inline source/System.cpp:972]: Functions should be defined in the same order as they are defined in the header file, which means this function shouldn't be below the price functions.

File: source/System.h [hunk 8]
```diff
@@ -137,6 +137,8 @@ class System {
 	bool HasShipyard() const;
 	// Check whether you can buy or sell ship outfits in this system.
 	bool HasOutfitter() const;
+	// Get the system's invisible fence radius.
+	double InvisibleFenceRadius() const;
```
Comment (Amazinite) [inline source/System.h:141]: I would move this up to above the jump range/below the asteroid belts functions.

File: source/System.cpp [hunk 9]
```diff
@@ -371,6 +371,8 @@ void System::Load(const DataNode &node, Set<Planet> &planets)
 					grand.PrintTrace("Warning: Skipping unsupported departure distance limitation:");
 			}
 		}
+		else if(key == "invisible fence" && child.Size() >= 2)
+			invisibleFenceRadius = child.Value(1);
```
Comment (Amazinite) [inline source/System.cpp:375]:
```suggestion
			invisibleFenceRadius = max(0., child.Value(1));
```

File: source/System.h [hunk 10]
```diff
@@ -234,6 +236,9 @@ class System {
 
 	// Attributes, for use in location filters.
 	std::set<std::string> attributes;
+
+	// The "invisible fence" radius for this system.
+	double invisibleFenceRadius = 10000.;
```
Comment (Amazinite) [inline source/System.h:241]: Same for the variable. Move it to between belts and jump range.

**Review by Amazinite — CHANGES_REQUESTED** [review]

File: source/System.h [hunk 11]
```diff
@@ -234,6 +240,7 @@ class System {
 
 	// Attributes, for use in location filters.
 	std::set<std::string> attributes;
+
```
Comment (Amazinite) [inline source/System.h:243]:
```suggestion
```

File: source/System.h (hunk 8 above)
Comment (Amazinite) [reply source/System.h:141]: It's not moved; :| :| :| :|

File: source/System.h [hunk 12]
```diff
@@ -209,7 +211,12 @@ class System {
 	std::vector<RandomEvent<Fleet>> fleets;
 	std::vector<RandomEvent<Hazard>> hazards;
 	double habitable = 1000.;
+
 	WeightedList<double> belts;
+
+	// The "invisible fence" radius for this system.
+	double invisibleFenceRadius = 10000.;
+
```
Comment (Amazinite) [inline source/System.h:219]:
```suggestion
	WeightedList<double> belts;
	// The "invisible fence" radius for this system.
	double invisibleFenceRadius = 10000.;
```

**Review by Amazinite — APPROVED** [review]

File: source/System.h [hunk 13]
```diff
@@ -210,6 +212,8 @@ class System {
 	std::vector<RandomEvent<Hazard>> hazards;
 	double habitable = 1000.;
 	WeightedList<double> belts;
+	// The "invisible fence" radius for this system.
```
Comment (Amazinite) [inline source/System.h:215]:
```suggestion
```

File: source/System.cpp [hunk 14]
```diff
@@ -686,6 +688,13 @@ const WeightedList<double> &System::AsteroidBelts() const
 
 
 
+double System::InvisibleFenceRadius() const
```
Comment (Amazinite) [inline source/System.cpp:691]:
```suggestion
// Get the system's invisible fence radius.
double System::InvisibleFenceRadius() const
```

**Pull request merged**